### DISCUSSION TITLE
How to remove Topolvm completely from a Kuberenetes cluster?

### DISCUSSION THREAD

**Discussion by saeedkaz** [topolvm/topolvm#644] [Q&A] [answered]
We had a Topolvm on our cluster. First I removed all data from it. After that, we tried to remove Topolvm from our cluster but it wasn't successful. I did the following steps First I reduce the replica set of Topolvm to 0 but it was unsuccessful and I get the following error.

MountVolume.SetUp failed for volume "kube-api-access-gvxr4" : object "namespace-test"/"kube-root-ca.crt" not registered 
After that, I deleted the webhook with no success. After that, I deleted the lines about Topolvm in the scheduler which make the cluster unstable.

I am looking for a way to cleanly and completely remove Topolvm from the cluster

**Comment by peng225** [comment]
HI @saeedkaz 

Sorry for the lack of the document about clean-up procedures.

Because TopoLVM has several settings for deployment, the clean-up procedure depends on them.
Could you tell me how you installed the TopoLVM on your Kubernetes cluster?
Especially, I would like to know your settings for the following parameter.
https://github.com/topolvm/topolvm/blob/main/charts/topolvm/values.yaml#L132

**Comment by saeedkaz** [comment]
Hi Shinya,
Thanks for you response.
I didn't know how TopoLVM was installed because it was installed by one of
our old colleagues.
The only thing that I know is that he used attached file for installation.

Regards,
Saeed

**Reply by peng225** [reply]
Hi @saeedkaz 

> I didn't know how TopoLVM was installed because it was installed by one of
our old colleagues.

O.K. Thanks.

Probably, you can clean up TopoLVM as follows.

1. Delete all PVC whose `.spec.storageClassName` is set to the TopoLVM's StorageClass.
    (You will also need to delete pods using those PVCs, if any.)
    Then, PV and LogicalVolume custom resources will be also deleted.
3. Delete the namespace where TopoLVM is deployed. (e.g. `topolvm-system`)
4. Delete StorageClasses whose provisioner is `topolvm.io` or (if running in legacy mode,) `topolvm.cybozu.com`.
5. If lvmd is running as a systemd unit, stop it.
    ```bash
    systemctl is-active lvmd.service
    systemctl stop lvmd.service
    ``` 
    (If lvmd was running as a DaemonSet, it has already been deleted in step 2.)

> The only thing that I know is that he used attached file for installation.

Sorry, but I couldn't find any attached files. Did you forget to attach it?

**Comment by saeedkaz** [comment]
Hi,
Thank you,
I have attached file again.
I've deleted all PVC's.
I can not delete namespace becasue it seems risky and it is a production
environment.
Regards,
Saeed

**Reply by peng225** [reply]
> I can not delete namespace becasue it seems risky and it is a production

O.K.
Unfortunately, I cannot answer right now about how to remove TopoLVM without deleting the namespace.
I created the following issue, and we will resolve it when we have time.
https://github.com/topolvm/topolvm/issues/647

If you still have questions after reading the document which will be written as an outpu of https://github.com/topolvm/topolvm/issues/647, please let us know again.

**Comment by peng225** [comment, marked as answer]
Hi @saeedkaz 

I discussed this topic with other maintainers, and they told me the way to clean up TopoLVM without deleting the namespace.

1. Delete TopoLVM's `MutatingWebhookConfiguration`. (The resource name is `topolvm-hook` by default.)
2. Edit the kube-scheduler's config file and remove TopoLVM's `extenders` settings. It may be your help to read [this configuration guide](https://github.com/topolvm/topolvm/blob/0ccaaa31779c2fd1c51f657a488e39d40afe5b46/deploy/README.md#configure-kube-scheduler).
   - Please be careful if you have any other settings in `extenders` section of the configuration file. You can remove only the TopoLVM's setting.
3. Delete TopoLVM's StorageClass.
4. Delete all other TopoLVM's resources.
    At least, you can delete all the following resources.
    (The resource name may be different in your environment. I show you the default name.)
    - deployment: `topolvm-controller`
    - daemonset: `topolvm-node`, `topolvm-scheduler`
       - There may also be lvmd here. It depends on your environment.
    - configmap: `topolvm-scheduler-options`
    - secret: `topolvm-mutatingwebhook`
5. Remove `topolvm.io/webhook=ignore` label from `kube-system` and `topolvm-system` namespaces.

About lvmd, please see the step 4 in [this comment](https://github.com/topolvm/topolvm/discussions/644#discussioncomment-4761922).

You may have installed cert-manager with TopoLVM. Whether you can delete cert-manager-related resources depends on your cluster. If you would like to delete them, please be careful if you can really delete them.
